File: app/src/main/java/org/autojs/autojs/ui/main/drawer/DrawerMenuAdapter.java
package org.autojs.autojs.ui.main.drawer;

import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.recyclerview.widget.RecyclerView;

import org.autojs.autojs.databinding.DrawerMenuGroupBinding;
import org.autojs.autojs.databinding.DrawerMenuItemBinding;
import org.autojs.autojs.ui.widget.BindableViewHolder;

import java.util.List;

/**
 * Created by Stardust on 2017/12/10.
 */

public class DrawerMenuAdapter extends RecyclerView.Adapter<BindableViewHolder<DrawerMenuItem>> {


    private static final int VIEW_TYPE_ITEM = 0;
    private static final int VIEW_TYPE_GROUP = 1;


    private final List<DrawerMenuItem> mDrawerMenuItems;

    public DrawerMenuAdapter(List<DrawerMenuItem> drawerMenuItems) {
        mDrawerMenuItems = drawerMenuItems;
    }

    public List<DrawerMenuItem> getDrawerMenuItems() {
        return mDrawerMenuItems;
    }

    @Override
    public BindableViewHolder<DrawerMenuItem> onCreateViewHolder(ViewGroup parent, int viewType) {
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        if (viewType == VIEW_TYPE_GROUP) {
            DrawerMenuGroupBinding binding = DrawerMenuGroupBinding.inflate(inflater, parent, false);
            return new DrawerMenuGroupViewHolder(binding);
        } else {
            DrawerMenuItemBinding binding = DrawerMenuItemBinding.inflate(inflater, parent, false);
            return new DrawerMenuItemViewHolder(binding);
        }
    }

    @Override
    public void onBindViewHolder(BindableViewHolder<DrawerMenuItem> holder, int position) {
        holder.bind(mDrawerMenuItems.get(position), position);
    }

    @Override
    public int getItemCount() {
        return mDrawerMenuItems.size();
    }

    @Override
    public int getItemViewType(int position) {
        return mDrawerMenuItems.get(position) instanceof DrawerMenuGroup ?
                VIEW_TYPE_GROUP : VIEW_TYPE_ITEM;
    }

    public void notifyItemChanged(DrawerMenuItem item) {
        int pos = mDrawerMenuItems.indexOf(item);
        notifyItemChanged(pos);
    }
}

File: app/src/main/java/org/autojs/autojs/ui/main/drawer/DrawerMenuAdapter.kt
package org.autojs.autojs.ui.main.drawer

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import org.autojs.autojs.databinding.DrawerMenuGroupBinding
import org.autojs.autojs.databinding.DrawerMenuItemBinding
import org.autojs.autojs.ui.widget.BindableViewHolder

class DrawerMenuAdapter(
    private val drawerMenuItems: List<DrawerMenuItem>
) : RecyclerView.Adapter<BindableViewHolder<DrawerMenuItem>>() {

    companion object {
        private const val VIEW_TYPE_ITEM = 0
        private const val VIEW_TYPE_GROUP = 1
    }

    fun getDrawerMenuItems(): List<DrawerMenuItem> = drawerMenuItems

    override fun onCreateViewHolder(
        parent: ViewGroup, viewType: Int
    ): BindableViewHolder<DrawerMenuItem> {
        val inflater = LayoutInflater.from(parent.context)
        return if (viewType == VIEW_TYPE_GROUP) {
            val binding = DrawerMenuGroupBinding.inflate(inflater, parent, false)
            DrawerMenuGroupViewHolder(binding)
        } else {
            val binding = DrawerMenuItemBinding.inflate(inflater, parent, false)
            DrawerMenuItemViewHolder(binding)
        }
    }

    override fun onBindViewHolder(holder: BindableViewHolder<DrawerMenuItem>, position: Int) {
        holder.bind(drawerMenuItems[position], position)
    }

    override fun getItemCount(): Int = drawerMenuItems.size

    override fun getItemViewType(position: Int): Int {
        return if (drawerMenuItems[position] is DrawerMenuGroup) VIEW_TYPE_GROUP else VIEW_TYPE_ITEM
    }

    fun notifyItemChanged(item: DrawerMenuItem) {
        val pos = drawerMenuItems.indexOf(item)
        notifyItemChanged(pos)
    }
}